### PR TITLE
Acessibility: added AccessibilityAction and SupportedAccessibilityAction

### PR DESCRIPTION
**Pull request by FloVanGH**
The adds only the enums / structs needed to implement accessibility actions. There is no functionality yet. This is the first step to implement #2872. The enum is a copy of the available accessibility actions provides by accesskit, before merging we need to check what is really needed for Slint and can be already implemented with the current feature set.

### CODE REVIEW

File: internal/core/accessibility.rs [hunk 1]
```diff
@@ -24,6 +30,68 @@ pub enum AccessibleStringProperty {
...
+    Blur,
+    Collapse,
+    Expand,
+    CustomAction,
+    Decrement,
+    Increment,
+    // FIXME: Can it be currently supported?
+    HideToolTip,
+    // FIXME: Can it be currently supported?
+    ShowToolTip,
```
Comment (ogoffart) [inline internal/core/accessibility.rs:46]: I guess there is no point having action that don't map to action that we can define in the .slint language. And we should only support what we can implement.

File: internal/core/accessibility.rs [hunk 2]
```diff
@@ -24,6 +30,68 @@ pub enum AccessibleStringProperty {
...
+    // FIXME: Can it be currently supported?
+    ShowContextMenu,
+}
+
+bitflags! {
+    // Define a accessibility actions that currently supported by Slint.
+    pub struct SupportedAccessibilityAction: u32 {
+        const Default = 0b00000001;
+        const Focus = 0b00000010;
+        const Blur = 0b00000011;
```
Comment (ogoffart) [inline internal/core/accessibility.rs:70]: I thought the const in a bitflag can only have a single bit set.

maybe the `1 << X` notation would be better

File: internal/core/accessibility.rs (hunk 1 above)
Comment (FloVanGH) [reply internal/core/accessibility.rs]: Agree. I have removed the unsupported ones.

File: internal/core/accessibility.rs (hunk 2 above)
Comment (FloVanGH) [reply internal/core/accessibility.rs]: I changed it, is that what you mean?

**Comment by tronical** [comment]
> ... before merging we need to check what is really needed for Slint and can be already implemented with the current feature set.

I share this sentiment. Perhaps we should just declare the ones we know are trivial to implement at this point and then add more later. That's basically Focus, Increment, Decrement, SetValue, SetTextSelection, SetValue, and perhaps ReplaceSelectedText. Maybe there are more?

File: internal/core/accessibility.rs [hunk 3]
```diff
@@ -24,6 +30,56 @@ pub enum AccessibleStringProperty {
...
+    ScrollBackward,
+    ScrollDown,
+    ScrollForward,
+    ScrollLeft,
+    ScrollRight,
+    ScrollUp,
+    ScrollIntoView,
+    ScrollToPoint(LogicalPoint),
+    SetScrollOffset(LogicalVector),
+    SetTextSelection(Option<core::ops::Range<i32>>),
```
Comment (tronical) [inline internal/core/accessibility.rs:53]: Are those byte offsets or char offsets?

File: internal/core/accessibility.rs [hunk 4]
```diff
@@ -24,6 +30,56 @@ pub enum AccessibleStringProperty {
     ValueStep,
 }
 
+// Defines an accessibility action.
+pub enum AccessibilityAction {
+    Default,
+    Focus,
+    Blur,
```
Comment (tronical) [inline internal/core/accessibility.rs:37]: I wonder what the expected behaviour here is. I wonder how to trigger this.

File: internal/core/accessibility.rs (hunk 3 above)
Comment (FloVanGH) [reply internal/core/accessibility.rs:53]: I checked how it is handled in `Flickable` and `TextInput` and used it here.

**Comment by FloVanGH** [comment]
> > ... before merging we need to check what is really needed for Slint and can be already implemented with the current feature set.
> 
> I share this sentiment. Perhaps we should just declare the ones we know are trivial to implement at this point and then add more later. That's basically Focus, Increment, Decrement, SetValue, SetTextSelection, SetValue, and perhaps ReplaceSelectedText. Maybe there are more?

Agree. Make it easier to implement and the whole thing. And additional features can be extended later.

File: internal/core/accessibility.rs (hunk 2 above)
Comment (FloVanGH) [reply internal/core/accessibility.rs]: @ogoffart ?

File: internal/core/accessibility.rs (hunk 3 above)
Comment (FloVanGH) [reply internal/core/accessibility.rs:40]: I think it is byte offset in `TextInput` right? So it is handled by the renderer.

File: internal/core/accessibility.rs (hunk 2 above)
Comment (ogoffart) [reply internal/core/accessibility.rs]: Yes, this is what i meant.

File: internal/core/accessibility.rs (hunk 3 above)
Comment (tronical) [reply internal/core/accessibility.rs:39]: 👍

**Comment by ogoffart** [comment]
I don't think we should merge this as is in master, because there is no use for it without the code that uses it.

**Comment by FloVanGH** [comment]
> I don't think we should merge this as is in master, because there is no use for it without the code that uses it.

Ok I can do a general wip/accessibility branch and we can merge it there. @ogoffart what do you think?

**Comment by tronical** [comment]
Good idea. If you call it `feature/accessibility` then it'll automatically CI tested (if you'd like that).

**Comment by FloVanGH** [comment]
> Good idea. If you call it `feature/accessibility` then it'll automatically CI tested (if you'd like that).

done. I have also change this PR from master to the feature/accessibility branch.

**Pull request merged**